Add support for an initial access token to registration endpoint

jogu opened the pull request:
As per the dynamic client registration specification, the AS may require an access token for access to the registration endpoint - allow the app to specify this.

(I've tried my best to follow the coding standard doc but it's quite long so very possible I missed something, let me know if so.)

WilliamDenniss commented:
Thanks for the contribution!

> As per the dynamic client registration specification

Can you add links in your method comments to the specific section of the specification that is implemented by your code? Search the codebase for `@see` for several examples of this.

Also for my own knowledge reviewing this, can you provide a summary here of how these access tokens are used during client registration?

Regarding coding standards, the style guide we follow is here: http://google.github.io/styleguide/objcguide.html

WilliamDenniss commented on Source/OIDRegistrationRequest.h:
```diff
@@ -79,7 +85,7 @@ NS_ASSUME_NONNULL_BEGIN
 - (instancetype)init NS_UNAVAILABLE;
 
-/*! @brief Designated initializer.
+/*! @brief Initializer for source compability with older versions
```
Is the intention to deprecate this method? If so it should be marked deprecated, if not then this wording should be changed.

jogu replied:
I'm not sure what the AppAuth policy on deprecations / source compatibility is. I guess perhaps deprecation should only happen at a major version release and actual API removal the following major version?

I've updated the text to say ' Initializer if no initial access token required'. I'm equally happy to mark it deprecated if you think that's better, or change it to a better wording. I'm 50:50.

jogu commented on Source/OIDRegistrationRequest.h:
```diff
@@ -34,6 +34,12 @@ NS_ASSUME_NONNULL_BEGIN
 @property(nonatomic, readonly) OIDServiceConfiguration *configuration;
 
+/*! @brief Any access token necessary to access the Client Registration Endpoint
+ @remarks OAuth 2.0 Access Token optionally issued by an Authorization Server granting
```
fixed

jogu commented on Source/OIDRegistrationRequest.m:
```diff
@@ -170,6 +201,10 @@ - (NSURLRequest *)URLRequest {
   URLRequest.HTTPMethod = kHTTPPost;
   [URLRequest setValue:kHTTPContentTypeHeaderValue forHTTPHeaderField:kHTTPContentTypeHeaderKey];
+  if (_initialAccessToken) {
+      NSString *value = [NSString stringWithFormat:@"%@ %@", kBearer, _initialAccessToken];
```
fixed

jogu commented on Source/OIDRegistrationRequest.h:
```diff
@@ -34,6 +34,12 @@ NS_ASSUME_NONNULL_BEGIN
 
+/*! @brief Any access token necessary to access the Client Registration Endpoint
+ @remarks OAuth 2.0 Access Token optionally issued by an Authorization Server granting
+        access to its Client Registration Endpoint
```
fixed.

jogu commented:
Thanks for the review! Apologies for the indent issues, this is the first iOS project I've worked on that uses 2 space indents. (I just put up a separate PR to specify that in the xcodeproj.)

> Can you add links in your method comments to the specific section of the specification that is implemented by your code? Search the codebase for @see for several examples of this.

Done.

> Also for my own knowledge reviewing this, can you provide a summary here of how these access tokens are used during client registration?

The spec ( https://openid.net/specs/openid-connect-registration-1_0.html#ClientRegistration ) says:

> "The OpenID Provider MAY require an Initial Access Token that is provisioned out-of-band (in a manner that is out of scope for this specification) to restrict registration requests to only authorized Clients or developers."

This is what we're doing on the project in question; to cut a long story short a form of authentication to the authorization server must happen before a client can be registered.

> Regarding coding standards, the style guide we follow is here: 
> http://google.github.io/styleguide/objcguide.html

Thanks, yes, I found it but given it's ~30 pages long I'm quite prepared to believe I've not yet fully understood all it's nuances. :-)

jogu commented:
Just pushed an update as I spotted another place (in the tests) where I'd got indentation wrong. I've addressed all your comments so hopefully this should be good to be merged now.

WilliamDenniss commented:
Thanks for the update, and the explanation! I agree it's in-scope. I'll review it again shortly now that I understand this part of the spec better (and with your recent changes).

jogu commented:
Thanks! I've just pushed a new version that applies all your suggested changes so hopefully good to go now.

WilliamDenniss commented on Source/OIDRegistrationRequest.h:
```diff
@@ -95,7 +106,29 @@ NS_ASSUME_NONNULL_BEGIN
+                          subjectType:(nullable NSString *)subjectType
+              tokenEndpointAuthMethod:(nullable NSString *)tokenEndpointAuthenticationMethod
+                   initialAccessToken:(nullable NSString *)initialAccessToken
+                 additionalParameters:(nullable NSDictionary<NSString *, NSString *> *)additionalParameters
```
line 131 is over 100 chars.

To resolve, the longest line of 125-131 (tokenEndpointAuthMethod) should be indented 4 spaces, and the rest lined up on that, as with the above initializer.

jogu commented:
That's fixed and ready to go. I also spotted that there was a different variable name in the header vs the previous initialiser so I updated the new one to name. (was tokenEndpointAuthenticationMethod, now tokenEndpointAuthMethod so it matches)
